platforms/jvm/testing-jvm-infrastructure/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestExecutionListener.java

```diff
@@ -365,21 +398,36 @@ private boolean isTestClassIdentifier(TestIdentifier testIdentifier) {
         return hasClassSource(testIdentifier) && hasDifferentSourceThanAncestor(testIdentifier);
     }
 
-    private static String className(@Nullable TestIdentifier testClassIdentifier) {
+    private String extractClassOrResourceName(TestIdentifier node) {
+        TestIdentifier testClassIdentifier = findTestClassIdentifier(node);
         if (testClassIdentifier != null) {
             Optional<ClassSource> classSource = getClassSource(testClassIdentifier);
             if (classSource.isPresent()) {
                 return classSource.get().getClassName();
             }
+        } else if (hasFileSource(node)) {
+            return computeRelativePathFromProject(node);
         }
-        return JUnitSupport.UNKNOWN_CLASS;
+
+        // Fall back to the unique id of the node.
+        // This prevents duplicate class names that our report can't handle,
+        // and provides appropriate information for non-class-based testing.
+        UniqueId.Segment lastSegment = getLastUniqueIdSegment(node);
+        // Remove ':' as we use them in Paths for reporting
+        return (lastSegment.getType() + "_" + lastSegment.getValue()).replace(':', '_');
     }
 
-    private static String classDisplayName(@Nullable TestIdentifier testClassIdentifier) {
-        if (testClassIdentifier != null) {
-            return testClassIdentifier.getDisplayName();
+    private String computeRelativePathFromProject(TestIdentifier node) {
+        try {
+            @SuppressWarnings("OptionalGetWithoutIsPresent")
+            FileSource fileSource = (FileSource) node.getSource().get();
+            Path rootDirPath = new File("").toPath().toRealPath();
```

Comment:
❓ Is the CWD of the worker reliably pointing to the project dir? This seems like an assumption we should not make, and instead explicitly pass in the project directory.